File: packages/cli/src/commands.ts
import { MagpieClient } from '@magpie/client';
import { normalizePairingCode } from '@magpie/protocol';
import { relayUrl, requireExtension } from './env.js';
import { writeSession, readSession, clearSession } from './store.js';
import { streamUntilDone } from './runtime.js';
import type { StreamResult } from './runtime.js';
import { saveReport, renderReport, listReports, readReport, outcomeLabel } from './reports.js';

/**
 * Print the end reason, then build + show + persist the call report. This is
 * the "report on termination" the async value proposition depends on: the
 * conclusion + transcript are saved to ~/.magpie/calls/ so an away human
 * can read them later with `magpie report`.
 */
function finishCall(
  client: MagpieClient,
  callId: string,
  result: StreamResult,
  io: Io,
): void {
  io.out(`\n${result.reason}`);
  const report = client.buildReport(callId, result.outcome);
  if (report) {
    io.out(renderReport(report));
    try {
      const path = saveReport(report);
      io.out(`📋 Report saved: ${path}  (re-read with: magpie report ${report.callId})`);
    } catch (err) {
      io.err(`(could not save report: ${err instanceof Error ? err.message : String(err)})`);
    }
  }
}

/**
 * The command handlers. Each is small and friendly: this is the
 * non-developer-simple surface, so output is minimal and reads like a phone
 * call, not a stack trace.
 *
 * IMPORTANT — process model: a call lives on the WebSocket connection that
 * opened or joined it (the relay keys calls by socket identity). So `start`,
 * `call`, and `listen` are LONG-LIVED: they hold the line open and stream
 * inbound queries until hangup. `join` patches in and then streams. `hangup`
 * is the only one-shot command — it signals the live process to drop the line.
 */

export interface Io {
  out: (line: string) => void;
  err: (line: string) => void;
  env: NodeJS.ProcessEnv;
  /** Line source for typing replies. Present on the real CLI; omitted in tests. */
  input?: NodeJS.ReadableStream;
}

export const consoleIo: Io = {
  out: (l) => process.stdout.write(`${l}\n`),
  err: (l) => process.stderr.write(`${l}\n`),
  env: process.env,
  input: process.stdin,
};

/** The shareable invite line a human pastes into chat. */
export function shareLine(code: string): string {
  return `Patch your agent in:  magpie join ${code}`;
}

/**
 * `start <topic>` — open a call, print the code + a shareable line, then hold
 * the line open and print any inbound queries. Does not block on the peer.
 */
export async function start(topic: string, io: Io = consoleIo): Promise<void> {
  const from = requireExtension(io.env);
  const url = relayUrl(io.env);
  const client = await MagpieClient.connect(url);
  const { code, callId } = await client.start({ from, topic });

  writeSession({
    code,
    callId,
    topic,
    from,
    role: 'opener',
    relayUrl: url,
    pid: process.pid,
    startedAt: new Date().toISOString(),
  });

  io.out(`☎️  Magpie line open for: ${topic}`);
  io.out('');
  io.out(`   Your code:  ${code}`);
  io.out(`   ${shareLine(code)}`);
  io.out('');
  io.out('Waiting on the line… (Ctrl-C to hang up)');

  const result = await streamUntilDone(client, io.out, {
    send: io.input ? { from, callId, input: io.input, peer: null } : undefined,
  });
  finishCall(client, callId, result, io);
  clearSession();
  client.close();
}

/**
 * `call <topic>` — like `start`, but waits for the peer to actually join before
 * announcing "patched through", then streams inbound queries.
 */
export async function call(topic: string, io: Io = consoleIo): Promise<void> {
  const from = requireExtension(io.env);
  const url = relayUrl(io.env);
  const client = await MagpieClient.connect(url);
  const { code, callId } = await client.start({ from, topic });

  writeSession({
    code,
    callId,
    topic,
    from,
    role: 'opener',
    relayUrl: url,
    pid: process.pid,
    startedAt: new Date().toISOString(),
  });

  io.out(`☎️  Calling about: ${topic}`);
  io.out('');
  io.out(`   Your code:  ${code}`);
  io.out(`   ${shareLine(code)}`);
  io.out('');
  io.out('Ringing… waiting for the other agent to pick up. (Ctrl-C to cancel)');

  // The client surfaces the peer joining via a message stream; the very first
  // inbound activity OR an explicit peer-joined is our "picked up" signal. We
  // print "patched through" on the first message and then keep streaming.
  let announced = false;
  const result = await streamUntilDone(client, io.out, {
    send: io.input ? { from, callId, input: io.input, peer: null } : undefined,
    onMessage: () => {
      if (!announced) {
        announced = true;
        io.out('✅ Patched through. The agents are talking.');
      }
    },
  });
  finishCall(client, callId, result, io);
  clearSession();
  client.close();
}

/**
 * `join <code>` — patch your agent into an existing call using a code shared
 * over chat, then hold the line and stream inbound queries.
 */
export async function join(rawCode: string, io: Io = consoleIo): Promise<void> {
  const from = requireExtension(io.env);
  const url = relayUrl(io.env);
  const code = normalizePairingCode(rawCode); // throws a friendly error on bad shape
  const client = await MagpieClient.connect(url);
  const { callId, peer } = await client.join({ from, code });

  writeSession({
    code,
    callId,
    topic: '(joined)',
    from,
    role: 'joiner',
    relayUrl: url,
    pid: process.pid,
    startedAt: new Date().toISOString(),
  });

  io.out(`✅ Patched through to ${peer}. You are on the line.`);
  io.out('Type a message + Enter to send; inbound queries appear below. (Ctrl-C to hang up)');

  const result = await streamUntilDone(client, io.out, {
    send: io.input ? { from, callId, input: io.input, peer } : undefined,
  });
  finishCall(client, callId, result, io);
  clearSession();
  client.close();
}

/**
 * `listen` — resume the active call (from the stored code) and print inbound
 * fenced queries. Use this when you started a call in another way and just
 * want to watch the line. It re-patches in as a participant using the stored
 * pairing code.
 */
export async function listen(io: Io = consoleIo): Promise<void> {
  const from = requireExtension(io.env);
  const session = readSession();
  if (!session) {
    throw new Error(
      'No active call to listen to. Start one with `magpie start "<topic>"`\n' +
        'or join one with `magpie join <code>`.',
    );
  }
  const url = relayUrl(io.env);
  const client = await MagpieClient.connect(url);
  const { callId } = await client.join({ from, code: session.code });

  io.out(`👂 Listening on the line for: ${session.topic}`);
  io.out('Inbound queries appear below, fenced as untrusted data. (Ctrl-C to stop)');

  const result = await streamUntilDone(client, io.out);
  finishCall(client, callId, result, io);
  client.close();
}

/**
 * `hangup` — drop the active call. Tells the relay (via a fresh connection on
 * the stored call) and signals the live `start`/`call`/`listen` process to
 * exit if one is running, then clears the stored session.
 */
export async function hangup(io: Io = consoleIo): Promise<void> {
  const session = readSession();
  if (!session) {
    io.out('Nothing to hang up — no active call.');
    return;
  }

  // Best effort: ask the relay to close the call so the peer is notified.
  try {
    const client = await MagpieClient.connect(session.relayUrl);
    await client.hangup(session.callId);
    client.close();
  } catch {
    // Relay unreachable; we still tear down locally below.
  }

  // Signal the long-lived process (if it is ours and still alive) to exit.
  if (session.pid && session.pid !== process.pid) {
    try {
      process.kill(session.pid, 'SIGINT');
    } catch {
      // process already gone
    }
  }

  clearSession();
  io.out('📴 Hung up.');
}

/**
 * `history` — list past calls and their outcomes (from ~/.magpie/calls/).
 * This is how an away human catches up on what their agent concluded.
 */
export async function history(io: Io = consoleIo): Promise<void> {
  const reports = listReports();
  if (reports.length === 0) {
    io.out('No past calls yet.');
    return;
  }
  io.out(`Past calls (${reports.length}):`);
  for (const r of reports) {
    io.out(`  ${r.endedAt}  ${outcomeLabel(r.outcome)}  with ${r.peer ?? '?'}  "${r.topic}"`);
    io.out(`      magpie report ${r.callId}`);
  }
}

/**
 * `report [callId]` — show a past call's report + full transcript. Defaults to
 * the most recent call.
 */
export async function showReport(callId: string | undefined, io: Io = consoleIo): Promise<void> {
  const report = callId ? readReport(callId) : listReports()[0];
  if (!report) {
    io.out(callId ? `No report for call ${callId}.` : 'No past calls yet.');
    return;
  }
  io.out(renderReport(report));
  io.out('\n──── transcript ────');
  if (report.transcript.length === 0) {
    io.out('(no messages)');
  }
  for (const e of report.transcript) {
    io.out(`[${e.type}] ${e.from}: ${e.content}`);
  }
}

File: rust/crates/magpie-cli/src/commands.rs
//! The command handlers.
//!
//! Mirror of `commands.ts`. Each is small and friendly — this is the
//! non-developer-simple surface, so output reads like a phone call. `start` and
//! `join` are LONG-LIVED: they hold the WebSocket open and stream inbound
//! queries until the call ends. `history` and `report` are one-shot reads of the
//! on-disk call store.

use magpie_client::{JoinOpts, StartOpts, MagpieClient};
use magpie_protocol::{format_invite, loopback_invite_warning, parse_invite};

use crate::env::{relay_url, require_extension};
use crate::reports::{
    list_reports, outcome_label, read_report, render_report, save_report,
};
use crate::runtime::{message_type_str, stream_until_done, SendOpts, StreamResult};

/// Errors surface as a single friendly `✗ …` line + exit 1 (see `main`).
type CmdResult = Result<(), Box<dyn std::error::Error + Send + Sync>>;

/// The shareable invite line a human pastes into chat. The invite token is
/// self-contained (`CODE@relay-url`), so the joiner needs no relay config.
/// Falls back to the bare code if the relay URL is not ws:// / wss:// —
/// a degraded-but-still-shareable line beats a hard failure here.
pub fn share_line(code: &str, relay_url: &str) -> String {
    let token = format_invite(code, relay_url).unwrap_or_else(|_| code.to_string());
    format!("Patch your agent in:  magpie join {token}")
}

/// Print the end reason, then build + show + persist the call report. This is
/// the "report on termination" the async value proposition depends on.
fn finish_call(client: &MagpieClient, call_id: &str, result: StreamResult) {
    println!("\n{}", result.reason);
    if let Some(report) = client.build_report(call_id, result.outcome) {
        println!("{}", render_report(&report));
        match save_report(&report) {
            Ok(path) => println!(
                "📋 Report saved: {}  (re-read with: magpie report {})",
                path.display(),
                report.call_id
            ),
            Err(e) => eprintln!("(could not save report: {e})"),
        }
    }
}

/// `start <topic>` — open a call, print the code + a shareable line, then hold
/// the line open and stream any inbound queries. Does not block on the peer.
pub async fn start(topic: &str) -> CmdResult {
    let from = require_extension()?;
    let url = relay_url();
    let client = MagpieClient::connect(&url).await?;
    let started = client
        .start(StartOpts {
            from: from.clone(),
            topic: topic.to_string(),
            max_turns: None,
        })
        .await?;

    println!("☎️  Magpie line open for: {topic}");
    println!();
    println!("   Your code:  {}", started.code);
    println!("   {}", share_line(&started.code, &url));
    if let Some(warning) = loopback_invite_warning(&url) {
        println!();
        println!("   {warning}");
    }
    println!();
    println!("Waiting on the line… (Ctrl-C to hang up)");

    let result = stream_until_done(
        &client,
        Some(SendOpts {
            from,
            call_id: started.call_id.clone(),
            peer: None,
        }),
    )
    .await;
    finish_call(&client, &started.call_id, result);
    client.close();
    Ok(())
}

/// `join <invite-or-code>` — patch your agent into an existing call using the
/// token shared over chat, then hold the line and stream inbound queries.
/// A full invite (`CODE@ws://relay`) carries the relay URL — zero config; a
/// bare code falls back to `MAGPIE_RELAY_URL` (or the localhost default).
pub async fn join(raw_code: &str) -> CmdResult {
    let from = require_extension()?;
    // Parse first for a friendly error on a bad-shaped code or relay URL.
    let invite = parse_invite(raw_code)?;
    let url = invite.relay_url.unwrap_or_else(relay_url);
    let code = invite.code;
    let client = MagpieClient::connect(&url).await?;
    let joined = client
        .join(JoinOpts {
            from: from.clone(),
            code,
        })
        .await?;

    println!("✅ Patched through to {}. You are on the line.", joined.peer);
    println!("Type a message + Enter to send; inbound queries appear below. (Ctrl-C to hang up)");

    let result = stream_until_done(
        &client,
        Some(SendOpts {
            from,
            call_id: joined.call_id.clone(),
            peer: Some(joined.peer.clone()),
        }),
    )
    .await;
    finish_call(&client, &joined.call_id, result);
    client.close();
    Ok(())
}

/// `history` — list past calls and their outcomes (from `~/.magpie/calls/`).
/// This is how an away human catches up on what their agent concluded.
pub fn history() {
    let reports = list_reports();
    if reports.is_empty() {
        println!("No past calls yet.");
        return;
    }
    println!("Past calls ({}):", reports.len());
    for r in &reports {
        println!(
            "  {}  {}  with {}  \"{}\"",
            r.ended_at,
            outcome_label(r.outcome),
            r.peer.as_deref().unwrap_or("?"),
            r.topic
        );
        println!("      magpie report {}", r.call_id);
    }
}

/// `report [callId]` — show a past call's report + full transcript. Defaults to
/// the most recent call.
pub fn show_report(call_id: Option<&str>) {
    let report = match call_id {
        Some(id) => read_report(id),
        None => list_reports().into_iter().next(),
    };
    let Some(report) = report else {
        match call_id {
            Some(id) => println!("No report for call {id}."),
            None => println!("No past calls yet."),
        }
        return;
    };

    println!("{}", render_report(&report));
    println!("\n──── transcript ────");
    if report.transcript.is_empty() {
        println!("(no messages)");
    }
    for e in &report.transcript {
        println!("[{}] {}: {}", message_type_str(e.r#type), e.from, e.content);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn share_line_is_a_copy_pasteable_self_contained_invite() {
        assert_eq!(
            share_line("K7F3-9M2P-XQ4R", "ws://192.168.0.13:8787"),
            "Patch your agent in:  magpie join K7F3-9M2P-XQ4R@ws://192.168.0.13:8787"
        );
    }

    #[test]
    fn share_line_falls_back_to_bare_code_on_non_ws_relay() {
        assert_eq!(
            share_line("K7F3-9M2P-XQ4R", "http://not-a-ws-relay"),
            "Patch your agent in:  magpie join K7F3-9M2P-XQ4R"
        );
    }
}
